File: hw5/python/triangulate_many.py
import numpy as np

def triangulate_many(xy1, xy2, P1, P2):
    """
    Arguments
        xy: Calibrated image coordinates in image 1 and 2
            [shape 3 x n]
        P:  Projection matrix for image 1 and 2
            [shape 3 x 4]
    Returns
        X:  Dehomogenized 3D points in world frame
            [shape 4 x n]
    """
    n = xy1.shape[1]
    X = np.zeros((4,n)) # Placeholder, replace with your implementation
    
    #SVD of A to get 3D world coordinates
    for i in range(n):
        x1 = xy1[0,i]
        y1 = xy1[1,i]
        x2 = xy2[0,i]
        y2 = xy2[1,i]
        
        
        A = np.array([x1*P1[2,:] - P1[0,:],
                      y1*P1[2,:] - P1[1,:],
                      x2*P2[2,:] - P2[0,:],
                      y2*P2[2,:] - P2[1,:]])
        U, R, V = np.linalg.svd(A)
        VT = V.T
        X[:,i] = VT[:,-1]
        
    # NORMALIZING
    X = X/X[-1]
    return X

File: Final project/matlab/triangulate_many.m
  
function X = triangulate_many(xy1, xy2, P1, P2)
    % Arguments
    %     xy: Calibrated image coordinates in image 1 and 2
    %         [shape 3 x n]
    %     P:  Projection matrix for image 1 and 2
    %         [shape 3 x 4]
    % Returns
    %     X:  Dehomogenized 3D points in world frame
    %         [shape 4 x n]

    n = size(xy1, 2);
    X = zeros(4, n); 
    A = zeros(4,4);
    for i = 1:n
        A(1,:) = xy1(2,i)*P1(3,:) - P1(2,:);
        A(2,:) = P1(1,:) - xy1(1,i)*P1(3,:);
        A(3,:) = xy2(2,i)*P2(3,:) - P2(2,:);
        A(4,:) = P2(1,:) - xy2(1,i)*P2(3,:);
%         A(2,:) = xy1(2,i)*P1(3,:) - P1(2,:);
%         A(1,:) = xy1(1,i)*P1(3,:) - P1(1,:);
%         A(4,:) = xy2(2,i)*P2(3,:) - P2(2,:);
%         A(3,:) = xy2(1,i)*P2(3,:) - P2(1,:);
        
        [~,~,V] = svd(A);
        X(:,i) = V(:,size(V,2));
        X(:,i) = X(:,i)./X(4,i);
        
    
    end
end